unix/modjni.c

```diff
@@ -0,0 +1,375 @@
+    }
+
+    nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError, "method not found"));
+}
+
+
+STATIC mp_obj_t jmethod_call(mp_obj_t self_in, mp_uint_t n_args, mp_uint_t n_kw, const mp_obj_t *args) {
+    if (n_kw != 0) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError, "kwargs not supported"));
+    }
+    mp_obj_jmethod_t *self = self_in;
+
+    const char *name = qstr_str(self->name);
+//    jstring meth_name = JJ(NewStringUTF, name);
+
+    jclass obj_class = JJ(GetObjectClass, self->obj);
+    jarray methods = JJ(CallObjectMethod, obj_class, Class_getMethods_mid);
+
+    return call_method(self->obj, name, methods, false, n_args, args);
+}
+
+STATIC const mp_obj_type_t jmethod_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_jmethod,
+    .print = jmethod_print,
+    .call = jmethod_call,
+//    .attr = jobject_attr,
+//    .locals_dict = (mp_obj_t)&jobject_locals_dict,
+};
+
+#ifdef __ANDROID__
+#define LIBJVM_SO "libdvm.so"
+#else
+#define LIBJVM_SO "libjvm.so"
+#endif
+
+STATIC void create_jvm() {
+    JavaVMInitArgs args;
+    JavaVMOption options;
+    options.optionString = "-Djava.class.path=.";
+    args.version = JNI_VERSION_1_6;
+    args.nOptions = 1;
+    args.options = &options;
+    args.ignoreUnrecognized = 0;
+
+    if (env) {
+        return;
+    }
+
+    void *libjvm = dlopen(LIBJVM_SO, RTLD_NOW | RTLD_GLOBAL);
+    if (!libjvm) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_OSError, "unable to load libjvm.so, use LD_LIBRARY_PATH"));
+    }
+    int (*_JNI_CreateJavaVM)(void*, void**, void*) = dlsym(libjvm, "JNI_CreateJavaVM");
+
+    int st = _JNI_CreateJavaVM(&jvm, (void**)&env, &args);
+    if (st < 0 || !env) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_OSError, "unable to create JVM"));
+    }
+
+    jclass class_class = JJ(FindClass, "java/lang/Class");
+    jclass method_class = JJ(FindClass, "java/lang/reflect/Method");
+    String_class = JJ(FindClass, "java/lang/String");
+
+    Class_getField_mid = (*env)->GetMethodID(env, class_class, "getField",
+                                     "(Ljava/lang/String;)Ljava/lang/reflect/Field;");
+    Class_getMethods_mid = (*env)->GetMethodID(env, class_class, "getMethods",
+                                     "()[Ljava/lang/reflect/Method;");
+    Class_getConstructors_mid = (*env)->GetMethodID(env, class_class, "getConstructors",
+                                     "()[Ljava/lang/reflect/Constructor;");
+    Method_getName_mid = (*env)->GetMethodID(env, method_class, "getName",
+                                     "()Ljava/lang/String;");
+    Method_toString_mid = (*env)->GetMethodID(env, method_class, "toString",
+                                     "()Ljava/lang/String;");
+}
+
+STATIC mp_obj_t mod_jni_cls(mp_obj_t cls_name_in) {
+    const char *cls_name = mp_obj_str_get_str(cls_name_in);
+    if (!env) {
+        create_jvm();
+    }
+    jclass cls = JJ(FindClass, cls_name);
```

Comment:
There's actually concept of local and global references, and rules how they should be used. But these rules appear to be written with Java -> Java native method invocation in mind, while here, it's  native C code -> Java invocation. And - everything works like this, without explicit dealing with global references, etc - both on desktop and Android. I may imagine it will be require to add them later (and then `__del__` finalizers with all the bloat they bring), but so far, so good.